storage/src/main/scala/coop/rchain/storage/Lmdb.scala

```diff
@@ -0,0 +1,1285 @@
+  assert(dirNameIn.isDefined && !dirNameIn.get.isEmpty)
+  assert(nameIn.isDefined && !nameIn.get.isEmpty)
+  assert(!baseDirIn.isDefined || !baseDirIn.get.isEmpty)
+  assert(Lmdb.minDbSize <= maxSizeIn && maxSizeIn <= Lmdb.maxDbSize)
+
+  val dirName =
+    if (dirNameIn.isDefined && !dirNameIn.get.isEmpty) dirNameIn.get
+    else { throw new RChainException("DB Directory is invalid") }
+  val dbName =
+    if (nameIn.isDefined && !nameIn.get.isEmpty) { nameIn.get } else {
+      throw new RChainException("Name is invalid")
+    }
+  val baseDir =
+    if (baseDirIn.isDefined) {
+      if (new File(baseDirIn.get).isDirectory) { baseDirIn.get } else {
+        throw new RChainException("Base Directory is invalid")
+      }
+    } else System.getProperty("user.dir") // not our problem if throws
+  val mapSizeIn: Long =
+    if (Lmdb.maxDbSize < maxSizeIn) {
+      Lmdb.maxDbSize
+    } else if (maxSizeIn < Lmdb.minDbSize) {
+      Lmdb.minDbSize
+    } else {
+      maxSizeIn
+    }
+
+  protected[storage] val dbDir = new File(baseDir.toString + "/" + dirName)
+  private var success = false
+  if (!dbDir.exists) {
+    success = dbDir.mkdir()
+    assert(success)
+  }
+
+  protected[storage] val dbDataFile = new File(dbDir.toPath + "/" + "data.mdb")
+  protected[storage] val dbLockFile = new File(dbDir.toPath + "/" + "lock.mdb")
+
+  val env: Env[ByteBuffer] =
+    if (isWritableIn)
+      create()
+        .setMaxDbs(1)
+        .setMapSize(mapSizeIn)
+        .open(dbDir, MDB_WRITEMAP)
+    else
+      create()
+        .setMaxDbs(1)
+        .setMapSize(mapSizeIn)
+        .open(dbDir)
+  val db: Dbi[ByteBuffer] =
+    if (isKeyToValuesIn) { env.openDbi(nameIn.get, MDB_CREATE, MDB_DUPSORT) } else {
+      env.openDbi(nameIn.get, MDB_CREATE)
+    }
+
+  // add a key-value row to the database
+  def put[K, V](key: K,
+                value: V,
+                txn: Option[Txn[ByteBuffer]] = None): Boolean = {
+    if (!isWritable) { return false }
+
+    if (key.isInstanceOf[String]
+        && getMaxKeySize < key.asInstanceOf[String].length) {
+      throw new RChainException("put(): key string is too long")
+    }
+    val bbKey: Option[ByteBuffer] = Bb.create(key)
+    val bbValue = Bb.create(value)
+    if (!bbKey.isDefined || !bbValue.isDefined) {
+      return false
+    }
+    putByteBuffer(bbKey.get, bbValue.get, txn)
+  }
+
+  def putByteBuffer(key: ByteBuffer,
+                    value: ByteBuffer,
+                    txnIn: Option[Txn[ByteBuffer]] = None): Boolean = {
+    if (!isWritable) { return false }
+
+    val txn =
+      if (txnIn == None) { env.txnWrite() } else { txnIn.get }
+    val cursor = db.openCursor(txn)
+    cursor.put(key, value)
+    cursor.close()
+    if (txnIn == None) { txn.commit() }
+    true
+  }
+
+  // The getXx[K](key:K) returns an array of Xs.
+  // This set of methods resist being expressed as a
+  // single generic method due to the lack of genericity
+  // of ByteBuffer.
+
+  def getInts[K](key: K,
```

Comment:
Comments aside, the only differences that I can see among these functions are:
1. the return type `Array[T]`, and
1. the getter function which produces a single value of type `T` from a cursor.
Is that right?

Might be simple to parameterize with something like
```
def getValues[K,T](keyIn: K, txnIn: Option[Txn[ByteBuffer]], getOne: (Cursor[ByteBuffer]) => T): Option[Array[T]]
```
and I'm sure our Scala experts could do even more good with that.

Reply:
Hard for me to check this with the PR split up into non-compiling units, but:
```
  def getInts[K](key: K, txnIn: Option[Txn[ByteBuffer]] = None): Option[Array[Int]] = {
    getValues(K, txnIn, _.`val`.getInt)
  }
```

Reply:
Sorry that the PR is split into non-compiling units.  I was asked to break the code submission into smaller pieces.

My first hope was that we could replace all the type specific methods
like getInts() and getStrings() with a single method getValues().
While this seems to be syntactically possible, I don't think it
simplifies the interface.  Let's start with Chris's suggestion:

def getValues[K,T](
  keyIn: K,
  txnIn: Option[Txn[ByteBuffer]],
  getValue: (Cursor[ByteBuffer]) => T)
: Option[Array[T]]

To call this method, the client would need to:
1. create a transaction
2. create a cursor from the transaction
3. call the method with a type specific method, for example:
     getValues(0, txn, getInt)

   where:
     def getInt(cursor: Cursor[ByteBuffer]): Int = { cursor.`val`.getInt }

If the client has to pass in a type specific method like getInt(),
then he is effectively calling a type specific method and may as well
just call the original getInts(0), which would be less error prone.

Since the cursor is generated from the transaction, we can simplify the
signature for getValues() in two ways:

One:

def getValues[K,T](
  keyIn: K,
  getValue: (Txn[ByteBuffer]) => T)
: Option[Array[T]]

This doesn't work because each call to getValue() would have to generate
a new cursor, which precludes the traversal of values in the cursor.

Two:

def getValues[K,T](
  keyIn: K,
  getValue: (Cursor[ByteBuffer]) => T)
: Option[Array[T]]

This has the same problem that the original suggestion had, which is that
the client would need to pass in a type specific getValue() method,
like getInt(), which defeats the simplification of a single method call
to get all value types.


Even if we keep the type specific calls like getInts() and getStrings()
there is the hope that we can abstract away some of the repetitive code
into the getValues() method.  For example, we can significantly shorten 
getInts(), but I can't get it to compile.
Search on "ERROR" to find the three errors.

  def getInts[K](key: K,
                 txnIn: Option[Txn[ByteBuffer]] = None)
  : Option[Array[Int]] = {
    val txn =
      if (txnIn == None) { env.txnRead() } else { txnIn.get }
    val cursor = db.openCursor(txn)

    try {

      return getValues(key, cursor, getInt)

    } catch {
      case e: RChainException =>
        Log("getInts(): " + e)
        return None
      case e: Throwable =>
        throw e
    } finally {
      cursor.close()
      if (txnIn == None) { txn.close() }
    }
    None
  }

  def getInt(cursor: Cursor[ByteBuffer]): Int = {
    cursor.`val`.getInt
  }


  def getValues[K,T](key: K,
                     cursor: Cursor[ByteBuffer],
                     getValue:(Cursor[ByteBuffer]) => T)
  : Option[Array[T]] = {

    if (!Lmdb.isStringOrPrimitive(key))
      throw new RChainException("getValues(): key is not primitive or string")

    var bbKey: Option[ByteBuffer] = None
    if (Lmdb.isPrimitive(key)) {
      bbKey = Bb.create(key)
    } else {
      bbKey = Some(Bb.strToBb(key.asInstanceOf[String]))
    }
    if (!bbKey.isDefined) {
      throw new RChainException("getValues(): key fails translation to Bb")
    }

    var outcome = cursor.get(bbKey.get, MDB_SET_KEY)
    if (!outcome) {
      return None
    }

    if (isKeyToValues) {
      val blobsBuf = new ArrayBuffer[T]()

      outcome = cursor.seek(MDB_FIRST_DUP)
      while (outcome) {
        blobsBuf += getValue(cursor)
        outcome = cursor.seek(MDB_NEXT_DUP)
      }
      if (blobsBuf.isEmpty) {
        return None
      }
// ERROR:
// Error:(193, 28) No ClassTag available for T
//    return Some(blobsBuf.toArray)
// ERROR:
// Error:(193, 28) not enough arguments for method toArray: (implicit evidence$1:
//    scala.reflect.ClassTag[T])Array[T].
//    Unspecified value parameter evidence$1.
//      return Some(blobsBuf.toArray)

      return Some(blobsBuf.toArray)
    } else {
      outcome = cursor.seek(MDB_GET_CURRENT)
      if (!outcome) {
        throw new RChainException("getValues(): MDB_GET_CURRENT")
      }

// ERROR:
// Error:(200, 24) cannot find class tag for element type T
//    val valueArray = new Array[T](1)

      val valueArray = new Array[T](1)
      valueArray(0) = getValue(cursor)
      return Some(valueArray)
    }
    None
  }